File: templates/cpp/src/mcap.cpp
#ifndef MCAP_COMPRESSION_NO_LZ4
#define MCAP_COMPRESSION_NO_LZ4
#endif
#ifndef MCAP_COMPRESSION_NO_ZSTD
#define MCAP_COMPRESSION_NO_ZSTD
#endif
#define MCAP_IMPLEMENTATION
#include <synapse/mcap.hpp>

#include <algorithm>
#include <cstring>
#include <stdexcept>
#include <unordered_map>

namespace synapse {
namespace mcap {

namespace {

::mcap::McapWriterOptions writerOptions(std::string_view library) {
  ::mcap::McapWriterOptions options(SYNAPSE_MCAP_PROFILE);
  options.library = std::string(library);
  options.noChunkCRC = true;
  options.enableDataCRC = false;
  options.noSummaryCRC = true;
  options.noChunking = true;
  options.noMessageIndex = true;
  options.noSummary = true;
  options.compression = ::mcap::Compression::None;
  options.noRepeatedSchemas = true;
  options.noRepeatedChannels = true;
  options.noAttachmentIndex = true;
  options.noMetadataIndex = true;
  options.noChunkIndex = true;
  options.noStatistics = true;
  options.noSummaryOffsets = true;
  return options;
}

void validateOpenArguments(std::string_view library,
                           std::string_view session_id,
                           std::string_view source) {
  if (library.empty()) throw std::invalid_argument("MCAP library identifier is empty");
  if (source.empty()) throw std::invalid_argument("Synapse MCAP source is empty");
  if (!Writer::validSessionId(session_id)) {
    throw std::invalid_argument(
        "Synapse MCAP session id must be 32 lowercase hexadecimal characters");
  }
}

}  // namespace

bool Writer::validSessionId(std::string_view value) {
  return value.size() == 32 &&
         std::all_of(value.begin(), value.end(), [](char value) {
           return (value >= '0' && value <= '9') ||
                  (value >= 'a' && value <= 'f');
         });
}

const char* Writer::timeBasisName(TimeBasis time_basis) {
  switch (time_basis) {
    case TimeBasis::MonotonicBoot:
      return SYNAPSE_MCAP_TIME_BASIS_MONOTONIC_BOOT;
    case TimeBasis::UnixEpoch:
      return SYNAPSE_MCAP_TIME_BASIS_UNIX_EPOCH;
    case TimeBasis::Correlated:
      return SYNAPSE_MCAP_TIME_BASIS_CORRELATED;
  }
  throw std::invalid_argument("unsupported Synapse MCAP time basis");
}

::mcap::Status Writer::open(std::ostream& output, std::string_view library,
                            std::string_view session_id,
                            std::string_view source, TimeBasis time_basis) {
  validateOpenArguments(library, session_id, source);
  writer_.open(output, writerOptions(library));
  return writeMetadata(session_id, source, time_basis);
}

::mcap::Status Writer::open(std::string_view filename,
                            std::string_view library,
                            std::string_view session_id,
                            std::string_view source, TimeBasis time_basis) {
  validateOpenArguments(library, session_id, source);
  auto status = writer_.open(filename, writerOptions(library));
  if (!status.ok()) return status;
  return writeMetadata(session_id, source, time_basis);
}

::mcap::Status Writer::writeMetadata(std::string_view session_id,
                                     std::string_view source,
                                     TimeBasis time_basis) {
  ::mcap::Metadata metadata;
  metadata.name = SYNAPSE_MCAP_METADATA_NAME;
  metadata.metadata = {
      {SYNAPSE_MCAP_SCHEMA_SET_HASH_KEY, SYNAPSE_SCHEMA_SET_HASH},
      {SYNAPSE_MCAP_SESSION_ID_KEY, std::string(session_id)},
      {SYNAPSE_MCAP_SOURCE_KEY, std::string(source)},
      {SYNAPSE_MCAP_TIME_BASIS_KEY, timeBasisName(time_basis)},
  };
  return writer_.write(metadata);
}

void Writer::addSchema(const Topic& topic, Schema& schema) {
  schema.topic = topic;
  schema.value.name = std::string(topic.name);
  schema.value.encoding = SYNAPSE_MCAP_SCHEMA_ENCODING;
  schema.value.data.assign(topic.schema_data, topic.schema_data + topic.schema_size);
  writer_.addSchema(schema.value);
}

void Writer::addChannel(const Schema& schema, std::string_view channel_topic,
                        Channel& channel) {
  channel.topic = schema.topic;
  channel.next_sequence = 0;
  channel.value = ::mcap::Channel(
      channel_topic, SYNAPSE_MCAP_MESSAGE_ENCODING, schema.value.id,
      {{SYNAPSE_MCAP_TOPIC_ID_KEY, std::to_string(schema.topic.id)}});
  writer_.addChannel(channel.value);
}

void Writer::addTopic(const Topic& topic, std::string_view channel_topic,
                      Channel& channel) {
  Schema schema;
  addSchema(topic, schema);
  addChannel(schema, channel_topic, channel);
}

::mcap::Status Writer::write(Channel& channel, std::uint64_t log_time_ns,
                             std::uint64_t publish_time_ns,
                             const std::byte* data, std::size_t size) {
  ::mcap::Message message;
  message.channelId = channel.value.id;
  message.sequence = channel.next_sequence++;
  message.logTime = log_time_ns;
  message.publishTime = publish_time_ns;
  message.data = data;
  message.dataSize = size;
  return writer_.write(message);
}

::mcap::Status Writer::writeFixed(Channel& channel,
                                  std::uint64_t log_time_ns,
                                  std::uint64_t publish_time_ns,
                                  const std::byte* payload, std::size_t size) {
  if (!channel.topic.fixed_layout) {
    throw std::invalid_argument(std::string(channel.topic.name) +
                                " is not fixed-layout");
  }
  if (size != channel.topic.payload_size) {
    throw std::invalid_argument("fixed payload size does not match topic catalog");
  }
  const std::size_t object_size = size + 4;
  if (object_size > UINT16_MAX || size > UINT32_MAX - 14) {
    throw std::length_error("fixed payload is too large to wrap");
  }
  fixed_buffer_.resize(size + 14);
  const std::uint32_t root = 4;
  const std::int32_t vtable_offset = -static_cast<std::int32_t>(object_size);
  const std::uint16_t vtable_size = 6;
  const std::uint16_t object_size_u16 = static_cast<std::uint16_t>(object_size);
  const std::uint16_t field_offset = 4;
  std::memcpy(fixed_buffer_.data(), &root, sizeof(root));
  std::memcpy(fixed_buffer_.data() + 4, &vtable_offset, sizeof(vtable_offset));
  std::memcpy(fixed_buffer_.data() + 8, payload, size);
  std::memcpy(fixed_buffer_.data() + 8 + size, &vtable_size, sizeof(vtable_size));
  std::memcpy(fixed_buffer_.data() + 10 + size, &object_size_u16,
              sizeof(object_size_u16));
  std::memcpy(fixed_buffer_.data() + 12 + size, &field_offset,
              sizeof(field_offset));
  return write(channel, log_time_ns, publish_time_ns, fixed_buffer_.data(),
               fixed_buffer_.size());
}

void Writer::close() { writer_.close(); }

::mcap::Status validateProfile(const ::mcap::McapReader& reader) {
  if (!reader.header().has_value()) {
    return {::mcap::StatusCode::InvalidFile, "missing MCAP header"};
  }
  if (reader.header()->profile != SYNAPSE_MCAP_PROFILE) {
    return {::mcap::StatusCode::InvalidFile,
            "unsupported MCAP profile: " + reader.header()->profile};
  }
  return {};
}

}  // namespace mcap
}  // namespace synapse

File: templates/python/mcap.py
"""First-class support for the frozen ``synapse/1`` MCAP profile.

Install with ``synapse-fbs[mcap]``. Container encoding and decoding are
delegated to the upstream :mod:`mcap` package; this module only applies the
Synapse profile, topic catalog, embedded BFBS, and fixed-struct wrapper.
"""

from dataclasses import dataclass
from enum import Enum
from importlib import resources
from typing import BinaryIO, Iterator, Union

from mcap.reader import make_reader
from mcap.writer import CompressionType, IndexType, Writer as ContainerWriter

from . import topic_catalog


class TimeBasis(str, Enum):
    MONOTONIC_BOOT = topic_catalog.MCAP_TIME_BASIS_MONOTONIC_BOOT
    UNIX_EPOCH = topic_catalog.MCAP_TIME_BASIS_UNIX_EPOCH
    CORRELATED = topic_catalog.MCAP_TIME_BASIS_CORRELATED


@dataclass
class TopicChannel:
    """A registered channel with per-channel sequence state."""

    id: int
    topic: topic_catalog.TopicInfo
    next_sequence: int = 0


def _valid_session_id(value: str) -> bool:
    return len(value) == 32 and all(char in "0123456789abcdef" for char in value)


def _topic(value: Union[str, topic_catalog.TopicInfo]) -> topic_catalog.TopicInfo:
    if isinstance(value, topic_catalog.TopicInfo):
        return value
    result = topic_catalog.topic_by_name(value)
    if result is None:
        raise KeyError(f"unknown Synapse topic: {value}")
    return result


def schema_data(topic: Union[str, topic_catalog.TopicInfo]) -> bytes:
    """Return the exact BFBS required by a topic's MCAP Schema record."""

    info = _topic(topic)
    relative = info.mcap_schema_file.removeprefix("bfbs/")
    return resources.files("synapse").joinpath("bfbs", relative).read_bytes()


def wrap_fixed_payload(payload: Union[bytes, bytearray, memoryview]) -> bytes:
    """Wrap fixed struct bytes in their existing one-field FlatBuffer root."""

    data = bytes(payload)
    object_size = len(data) + 4
    if object_size > 0xFFFF or len(data) > 0xFFFFFFFF - 14:
        raise ValueError("fixed payload is too large to wrap")
    return b"".join(
        (
            (4).to_bytes(4, "little"),
            (-object_size).to_bytes(4, "little", signed=True),
            data,
            (6).to_bytes(2, "little"),
            object_size.to_bytes(2, "little"),
            (4).to_bytes(2, "little"),
        )
    )


class Writer:
    """Uncompressed, unchunked, index-less ``synapse/1`` writer."""

    def __init__(
        self,
        output: BinaryIO,
        library: str,
        session_id: str,
        source: str,
        time_basis: TimeBasis = TimeBasis.MONOTONIC_BOOT,
    ) -> None:
        if not library:
            raise ValueError("MCAP library identifier is empty")
        if not source:
            raise ValueError("Synapse MCAP source is empty")
        if not _valid_session_id(session_id):
            raise ValueError(
                "Synapse MCAP session id must be 32 lowercase hexadecimal characters"
            )
        self._writer = ContainerWriter(
            output,
            compression=CompressionType.NONE,
            index_types=IndexType.NONE,
            repeat_channels=False,
            repeat_schemas=False,
            use_chunking=False,
            use_statistics=False,
            use_summary_offsets=False,
            enable_crcs=False,
            enable_data_crcs=False,
        )
        self._writer.start(profile=topic_catalog.MCAP_PROFILE, library=library)
        self._writer.add_metadata(
            topic_catalog.MCAP_METADATA_NAME,
            {
                topic_catalog.MCAP_SCHEMA_SET_HASH_KEY: topic_catalog.SCHEMA_SET_HASH,
                topic_catalog.MCAP_SESSION_ID_KEY: session_id,
                topic_catalog.MCAP_SOURCE_KEY: source,
                topic_catalog.MCAP_TIME_BASIS_KEY: TimeBasis(time_basis).value,
            },
        )

    def add_topic(
        self,
        topic: Union[str, topic_catalog.TopicInfo],
        channel_topic: str,
    ) -> TopicChannel:
        info = _topic(topic)
        schema_id = self._writer.register_schema(
            name=info.mcap_schema_name,
            encoding=topic_catalog.MCAP_SCHEMA_ENCODING,
            data=schema_data(info),
        )
        channel_id = self._writer.register_channel(
            topic=channel_topic,
            message_encoding=topic_catalog.MCAP_MESSAGE_ENCODING,
            schema_id=schema_id,
            metadata={topic_catalog.MCAP_TOPIC_ID_KEY: str(info.id)},
        )
        return TopicChannel(channel_id, info)

    def write(
        self,
        channel: TopicChannel,
        log_time_ns: int,
        publish_time_ns: int,
        data: Union[bytes, bytearray, memoryview],
    ) -> None:
        self._writer.add_message(
            channel_id=channel.id,
            sequence=channel.next_sequence,
            log_time=log_time_ns,
            publish_time=publish_time_ns,
            data=bytes(data),
        )
        channel.next_sequence = (channel.next_sequence + 1) & 0xFFFFFFFF

    def write_fixed(
        self,
        channel: TopicChannel,
        log_time_ns: int,
        publish_time_ns: int,
        payload: Union[bytes, bytearray, memoryview],
    ) -> None:
        if not channel.topic.fixed_layout:
            raise ValueError(f"{channel.topic.name} is not fixed-layout")
        if channel.topic.payload_size != len(payload):
            raise ValueError(
                f"fixed payload is {len(payload)} bytes, "
                f"expected {channel.topic.payload_size}"
            )
        self.write(
            channel,
            log_time_ns,
            publish_time_ns,
            wrap_fixed_payload(payload),
        )

    def finish(self) -> None:
        self._writer.finish()


class Reader:
    """Validated ``synapse/1`` reader backed by the upstream MCAP reader."""

    def __init__(self, stream: BinaryIO) -> None:
        self._reader = make_reader(stream)
        header = self._reader.get_header()
        if header.profile != topic_catalog.MCAP_PROFILE:
            raise ValueError(f"unsupported MCAP profile: {header.profile!r}")
        metadata = next(
            (
                item
                for item in self._reader.iter_metadata()
                if item.name == topic_catalog.MCAP_METADATA_NAME
            ),
            None,
        )
        if metadata is None:
            raise ValueError("missing required Synapse MCAP metadata")
        schema_set_hash = metadata.metadata.get(topic_catalog.MCAP_SCHEMA_SET_HASH_KEY, "")
        if len(schema_set_hash) != 32 or any(
            char not in "0123456789abcdef" for char in schema_set_hash
        ):
            raise ValueError("invalid Synapse MCAP schema-set hash")
        if not _valid_session_id(
            metadata.metadata.get(topic_catalog.MCAP_SESSION_ID_KEY, "")
        ):
            raise ValueError("invalid Synapse MCAP session id")
        if not metadata.metadata.get(topic_catalog.MCAP_SOURCE_KEY):
            raise ValueError("missing Synapse MCAP source")
        if metadata.metadata.get(topic_catalog.MCAP_TIME_BASIS_KEY) not in {
            basis.value for basis in TimeBasis
        }:
            raise ValueError("invalid Synapse MCAP time basis")
        self.metadata = metadata.metadata

    def messages(self, **kwargs: object) -> Iterator[object]:
        """Yield upstream ``(schema, channel, message)`` tuples."""

        yield from self._reader.iter_messages(**kwargs)


__all__ = [
    "Reader",
    "TimeBasis",
    "TopicChannel",
    "Writer",
    "schema_data",
    "wrap_fixed_payload",
]
